contracts/src/Teleporter/upgrades/TeleporterRegistry.sol

```diff
@@ -0,0 +1,69 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "../../WarpProtocolRegistry.sol";
+import "../ITeleporterMessenger.sol";
+
+/**
+ * @dev TeleporterRegistry contract is a {WarpProtocolRegistry} and provides an upgrade
+ * mechanism for teleporter messenger contracts.
+ */
+contract TeleporterRegistry is WarpProtocolRegistry {
+    mapping(address => uint256) private _addressToVersion;
+
+    constructor(
+        uint256[] memory initialVersions,
+        address[] memory initialProtocolAddresses
+    ) WarpProtocolRegistry(initialVersions, initialProtocolAddresses) {}
+
+    /**
+     * @dev Gets the {ITeleporterMessenger} contract of the given `version`.
+     */
+    function getTeleporterFromVersion(
+        uint256 version
+    ) external view returns (ITeleporterMessenger) {
+        return ITeleporterMessenger(_getAddressFromVersion(version));
+    }
+
+    /**
+     * @dev Gets the latest {ITeleporterMessenger} contract.
+     */
+    function getLatestTeleporter()
+        external
+        view
+        returns (ITeleporterMessenger)
+    {
+        return ITeleporterMessenger(_getAddressFromVersion(_latestVersion));
+    }
+
+    /**
+     * @dev Gets the version of the given `protocolAddress`.
+     * If `protocolAddress` is not a valid protocol address, returns 0.
+     */
+    function getVersionFromAddress(
+        address protocolAddress
+    ) external view returns (uint256) {
+        return _addressToVersion[protocolAddress];
+    }
+
+    /**
+     * @dev See {WarpProtocolRegistry-_addToRegistry}
+     *
+     * Adds the new protocol version address to the `_addressToVersion` mapping
+     * so that the registry can be queried for the version of a given protocol address.
+     * Requirements:
+     *
+     * - `version` must be the increment of the latest version.
```

Comment:
Do we need to verify this requirement? Can we determine what the "next" version should be, and if we can, perhaps we shouldn't take an arbitrary version as a parameter.

Reply:
The `version` is also acting as a "nonce", without it someone can take an old signed and valid warp out-of-band message then resubmit it to the registry, and it'd be valid through the warp precompile. 

@michaelkaplan13 can you double check my understanding here? Will look to make more clear in documentation.

Reply:
I think we'd probably be okay with just a sparse map. Being able to include add version that we don't already have. 

I don't think we have to worry about replays of old out-of-band messages. Warp messages come with a destination address and destination chain ID, so we could only ever deliver messages meant for this contract.

Reply:
The worry is for a replay of a warp out of band message that was sent on this same chain and registry address. i.e. A valid warp message registers address A at version 1, someone can send that same Warp message again to the registry, and it'd be valid unless it includes the version.

Reply:
Yeah, we definitely need to include the version with the warp message. Otherwise there is a risk of registries getting out of sync.

Reply:
This makes sense to me. I think I understand why the "version" must be included. Thank you for the explanation.